### PR TITLE
Add deep_merge support. Fixes GH-38 and GH-62

### PR DESCRIPTION
**Pull request by jyaworski**
Feedback welcome.

### CODE REVIEW

**Comment by jyaworski** [comment]
Tag @igalic. You're welcome. :)

File: manifests/deep_merge.pp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+#
+# This class installs and configures deep_merge
+#
+# === Authors:
+#
+# Terri Haber <terri@puppetlabs.com>
+#
+# === Copyright:
+#
+# Copyright (C) 2014 Terri Haber, unless otherwise noted.
```
Comment (igalic) [inline manifests/deep_merge.pp]: it's cool that you're attributing this to @terrimonster.

File: manifests/init.pp [hunk 2]
```diff
@@ -75,6 +75,9 @@
     unless $merge_behavior in ['native', 'deep', 'deeper'] {
       fail("${merge_behavior} merge behavior is invalid. Valid values are: native, deep, deeper")
     }
+    if $merge_behavior == 'deeper' {
```
Comment (igalic) [inline manifests/init.pp]: != native ?

Comment (jyaworski) [inline manifests/init.pp]: My thinking was that the docs discourage 'deep,' but I'll change it.

Comment (igalic) [inline manifests/init.pp]: they do, but we should still do the right thing™

File: manifests/deep_merge.pp (hunk 1 above)
Comment (jyaworski) [inline manifests/deep_merge.pp]: I copied most of the code. :)

Comment (terrimonster) [inline manifests/deep_merge.pp]: I'd appreciate you changing the attribution to yourself. Thank you!

**Comment by jyaworski** [comment]
OK, ready to merge.

**Comment by vchepkov** [comment]
Hi, I was about to submit something similar. It looks like most of the arguments you copied from eyaml class are unnecessary. all you need is $provider, $deep_merge_version and $gem_source

**Comment by vchepkov** [comment]
One more thing, you should not use 'notify', because it would break bootstrap scripts which are running puppet apply

I submitted pull requests to address it #106

**Comment by jyaworski** [comment]
Thanks @vchepkov. I've updated the PR with your suggestions. Do you have any info on why notify shouldn't be used? This is the first I'm hearing about it actually.

**Comment by vchepkov** [comment]
Sure.
Adding notify => Service would require for Service already exists in the catalog, otherwise it would fail to compile. In most cases you don't define puppet master service in your code, it comes with the product, so in order it to be in the catalog it has to be run via puppet agent call only.
it works well when your puppet master is already running and configured to apply your code, but for the fresh installation you have to run 'puppet apply' for some 'bootstrap' code and it would fail if you have this dependency. I didn't invent this idea, I 'borrowed' it from zack-r10k code, which is also needed for 'bootstrapping'. Hopefully this helps

**Comment by jyaworski** [comment]
That's really slick, what you did in #106. Using it that way doesn't presuppose the existence of the service.

+1

**Comment by smbambling** [comment]
would I also suggest that an additional parameter be added to this PR to allow you to sett deep_merge_options ?

Currently I'm using the extra_config options parameter in my profile..but that seems 'hacky' and seems like it was just a catch  at the time.

```
$hiera_extra_config =  '
:deep_merge_options:
  :knockout_prefix: \'--\''

class { 'hiera':
  extra_config         => $hiera_extra_config,
}
```

**Comment by jyaworski** [comment]
@smbambling sure. I'll get on that. Is there a handy list of deep merge options?

**Comment by smbambling** [comment]
@jyaworski:  Here is the doc for options with deep_merge https://github.com/danielsdeleo/deep_merge#options

Though I think something more generic that doesn't limit to the current options might be best for future proofing

**Comment by jyaworski** [comment]
@smbambling how's that?

**Comment by smbambling** [comment]
@jyaworski thats exactly what I was thinking :) :+1:

**Comment by jyaworski** [comment]
Ping @hunner. This is ready for merging.

File: manifests/init.pp [hunk 3]
```diff
@@ -82,6 +85,9 @@
     unless $merge_behavior in ['native', 'deep', 'deeper'] {
       fail("${merge_behavior} merge behavior is invalid. Valid values are: native, deep, deeper")
     }
+    if $merge_behavior != 'native' {
```
Comment (hunner) [inline manifests/init.pp]: Should this instead be `== 'deep'` or is it native for a reason?

Comment (jyaworski) [inline manifests/init.pp]: Anything but native needs the gem.

**Pull request merged**